tutorial.cxx

```diff
@@ -653,7 +654,81 @@ void f32_matrix_multiplication_4x4_loop_avx512_kernel(float a[4][4], float b[4][
 
     _mm512_storeu_ps(&c[0][0], c_mat);
 }
+*/
 
+void f32_matrix_multiplication_4x4_loop_avx512_kernel(float a[4][4], float b[4][4], float c[4][4]) {
+    __m512 a_mat = _mm512_loadu_ps(&a[0][0]);
+    __m512 b_mat = _mm512_loadu_ps(&b[0][0]);
+
+    __m512i trans_perm = _mm512_setr_epi32( //
+        0, 4, 8, 12,                        //
+        1, 5, 9, 13,                        //
+        2, 6, 10, 14,                       //
+        3, 7, 11, 15                        //
+    );
+    
+    __m512i lshiftz_8_perm = _mm512_setr_epi32( //
+        8, 9, 10, 11,                           //
+        12, 13, 14, 15,                         //
+        0, 0, 0, 0,                             //
+        0, 0, 0, 0                              //
+    );
+    
+    __m512i rot_4_perm = _mm512_setr_epi32( //
+        4, 5, 6, 7,                         //
+        8, 9, 10, 11,                       //
+        12, 13, 14, 15,                     //
+        0, 1, 2, 3                          //
```

Comment:
I assume the entries {0, 1, 2, 3} aren't used at all, and we can use the 255, or some other high integer to signal that.